fix log errors in tests

handler functions call metrics, which wont work because the config is not there. need to refactor metrics to properly test. 

```
Testing local command execution
Testing validate display on conversion
Testing validate display off conversion
Testing validate service actions
Testing GET /api/alive
2019/01/24 23:05:40 [ERROR] metric : open /etc/api/api_config.json: no such file or directory
PASS
ok      github.com/rebelit/rpIoT        0.016s
```

--- a/DockerFiles/test.sh
+++ b/DockerFiles/test.sh
@@ -1,5 +1,7 @@
 #!/bin/bash
 echo "Start: testing the program"
+mkdir -p /etc/api/
+cp /go/src/github.com/rebelit/rpIoT/api_config.json /etc/api/api_config.json
 cd /go/src/github.com/rebelit/rpIoT
 go test
 echo "Done: building the program"

--- a/README.md
+++ b/README.md
@@ -28,12 +28,12 @@ locally!
 * Build the image:
     * `docker build -t rpiot .`
 * Run all Tests:
-    * `docker run -v $PWD:/go/src/github.com/rebelit/rpIoT -i -t --rm rpiottest '/test.sh'`
+    * `docker run -v $PWD:/go/src/github.com/rebelit/rpIoT -i -t --rm rpiot /test.sh`
 * Build the program:
-    * `docker run -v $PWD:/go/src/github.com/rebelit/rpIoT -i -t --rm rpiottest '/build.sh'`
+    * `docker run -v $PWD:/go/src/github.com/rebelit/rpIoT -i -t --rm rpiot /build.sh`
     * `scp main pi@yourHost:/your/directory/here/`
 * Run the program:
-    * `docker run -v $PWD:/go/src/github.com/rebelit/rpIoT -i -t --rm rpiottest '/run.sh'`
+    * `docker run -v $PWD:/go/src/github.com/rebelit/rpIoT -i -t --rm rpiot /run.sh`
     
 
 ## Using it:

--- a/common.go
+++ b/common.go
@@ -9,7 +9,6 @@ import (
 	"gopkg.in/alexcesaro/statsd.v2"
 	"io/ioutil"
 	"log"
-	"os"
 	"os/exec"
 	"strconv"
 )
@@ -84,22 +83,18 @@ func readConfig()(Config, error){
 	return c, nil
 }
 
-func sendMetric(uri string, responseCode int, method string ){
+func sendMetric(uri string, responseCode int, method string ) error{
 	c, err := readConfig()
 	if err != nil{
 		log.Printf("[ERROR] metric : %s", err)
-		return
+		return err
 	}
-	host, err := os.Hostname()
-	if err != nil{
-		host = "unknown"
-	}
-	measurement := host+"_web-api"
+	measurement := "rpiot_web_api"
 	if c.Statsd_host == "" {
 		//no statsd host configured, log metric and move along...move along...
 		log.Printf("[INFO] metric : %s,uri_path=%s,response_code=%s,method=%s", measurement, uri,
 			strconv.Itoa(responseCode),method)
-		return
+		return err
 	}
 
 	tags := statsd.Tags("uri_path", uri, "response_code", strconv.Itoa(responseCode), "method", method)
@@ -112,7 +107,7 @@ func sendMetric(uri string, responseCode int, method string ){
 	defer client.Close()
 
 	client.Increment(measurement)
-	return
+	return err
 }
 
 func strToUint8(number string)(uint8, error){

--- a/common_test.go
+++ b/common_test.go
@@ -6,7 +6,7 @@ import (
 )
 
 func TestCommand(t *testing.T){
-	fmt.Println("Testing local command execution")
+	fmt.Println("***Testing local command execution***")
 	testCmd := "ls"
 	testArgs := []string{"$home"}
 
@@ -17,40 +17,60 @@ func TestCommand(t *testing.T){
 }
 
 func TestValidateDisplayActionOn(t *testing.T){
-	fmt.Println("Testing validate display on conversion")
-	testState := "on"
+	fmt.Println("***Testing validate display on conversion***")
+	testState := []string{"on", "off"}
 
-	val, err := validateDisplayAction(testState)
-	if err != nil{
-		t.Error(err)
-	}
-
-	if val != "1"{
-		t.Errorf("expected 1 got %s", val)
-	}
-}
-
-func TestValidateDisplayActionOff(t *testing.T){
-	fmt.Println("Testing validate display off conversion")
-	testState := "off"
-
-	val, err := validateDisplayAction(testState)
-	if err != nil{
-		t.Error(err)
-	}
-
-	if val != "0"{
-		t.Errorf("expected 0 got %s", val)
+	for _, s := range testState{
+		val, err := validateDisplayAction(s)
+		if err != nil{
+			t.Error(err)
+		}
+		if s == "on" {
+			if val != "1" {
+				t.Errorf("expected 1 got %s", val)
+			}
+		}else{
+			if val != "0" {
+				t.Errorf("expected 1 got %s", val)
+			}
+		}
 	}
 }
 
 func TestValidateServiceActions(t *testing.T){
-	fmt.Println("Testing validate service actions")
+	fmt.Println("***Testing validate service actions***")
 	testStates := []string{"start", "stop", "restart"}
 
 	for _, state := range testStates{
 		if err := validateServiceAction(state); err != nil{
 			t.Error(err)
 		}
+	}
+}
+
+func TestCheckConfig(t *testing.T){
+	fmt.Println("***Testing reading config file***")
+	//check for a value I know exists in the default docker testing setup
+	ok := "reboot"
+	val, err := checkConfig(ok)
+	if err != nil{
+		t.Error(err)
+	}
+	if !val{
+		t.Error("value should have been true")
+	}
+
+	//check for a value i knoe is not present in the default docker testing setup
+	bad := "dafuk"
+	val, err = checkConfig(bad)
+	if err == nil{
+		t.Errorf("This should have errored out because the value does not exist in the config %s, %v", bad, val)
+	}
+}
+
+func TestSendingMetric(t *testing.T){
+	fmt.Println("***Testing metric formatting***")
+	if err := sendMetric("/bunk", 200, "POST"); err != nil{
+		t.Error(err)
 	}
 }

--- a/httpHandlers.go
+++ b/httpHandlers.go
@@ -3,10 +3,52 @@ package main
 import (
 	"encoding/json"
 	"github.com/gorilla/mux"
+	"github.com/pkg/errors"
 	"log"
 	"net/http"
 	"strconv"
 )
+
+//Http Response helper functions
+func returnOk(w http.ResponseWriter, r *http.Request, resp Response){
+	code := http.StatusOK
+	sendMetric(r.URL.Path, code, r.Method)
+
+	w.Header().Set("Content-Type", "application/json; charset=UTF-8")
+	w.WriteHeader(code)
+
+	if err := json.NewEncoder(w).Encode(resp); err != nil {
+		log.Printf("[ERROR] %s : %s\n", r.URL.Path, err)
+	}
+}
+
+func returnBad(w http.ResponseWriter, r *http.Request, resp Response, err error){
+	code := http.StatusBadRequest
+	sendMetric(r.URL.Path, code, r.Method)
+
+	resp.Message = err.Error()
+
+	w.Header().Set("Content-Type", "application/json; charset=UTF-8")
+	w.WriteHeader(code)
+
+	if err := json.NewEncoder(w).Encode(resp); err != nil {
+		log.Printf("[ERROR] %s : %s\n", r.URL.Path, err)
+	}
+}
+
+func returnInternalError(w http.ResponseWriter, r *http.Request, resp Response, err error){
+	code := http.StatusInternalServerError
+	sendMetric(r.URL.Path, code, r.Method)
+
+	resp.Message = err.Error()
+
+	w.Header().Set("Content-Type", "application/json; charset=UTF-8")
+	w.WriteHeader(code)
+
+	if err := json.NewEncoder(w).Encode(resp); err != nil {
+		log.Printf("[ERROR] %s : %s\n", r.URL.Path, err)
+	}
+}
 
 //Namespace handlers
 func getStatus (w http.ResponseWriter, r *http.Request){
@@ -14,94 +56,61 @@ func getStatus (w http.ResponseWriter, r *http.Request){
 	resp.Namespace = r.URL.Path
 	resp.Message = "alive"
 
-	w.Header().Set("Content-Type", "application/json; charset=UTF-8")
-
-	if err := json.NewEncoder(w).Encode(resp); err != nil {
-		log.Printf("[ERROR] %s : %s\n", r.URL.Path, err)
-	}
-	sendMetric(r.URL.Path,http.StatusOK, r.Method)
+	returnOk(w,r,resp)
 }
 
 func powerAction (w http.ResponseWriter, r *http.Request) {
 	vars := mux.Vars(r)
 	action := vars["action"]
 
+	resp := Response{}
+	resp.Namespace = string(r.URL.Path)
+
 	enabled, err := checkConfig(action)
 	if err != nil{
-		log.Printf("[ERROR] %s : %s\n", r.URL.Path, err)
-		w.WriteHeader(http.StatusNotFound)
-		sendMetric(r.URL.Path,http.StatusNotFound, r.Method)
+		returnBad(w,r,resp, err)
 		return
 	}
 
 	if !enabled{
-		log.Printf("[WARN] %s : %s", r.URL.Path, "API function disabled")
-		w.WriteHeader(http.StatusBadRequest)
-		sendMetric(r.URL.Path,http.StatusBadRequest, r.Method)
+		returnBad(w,r,resp, errors.New("API function disabled"))
 		return
 	}
 
 	if err := command(string(action), []string{"+1"}); err != nil{
-		log.Printf("[ERROR] %s : %s", r.URL.Path, err)
-		w.WriteHeader(http.StatusInternalServerError)
-		sendMetric(r.URL.Path,http.StatusInternalServerError, r.Method)
+		returnInternalError(w,r,resp, err)
 		return
 	}
 
-	outputs := Response{}
-	outputs.Namespace = string(r.URL.Path)
-	outputs.Message = "success"
-
-	w.Header().Set("Content-Type", "application/json; charset=UTF-8")
-
-	if err := json.NewEncoder(w).Encode(outputs); err != nil {
-		log.Printf("[ERROR] %s : %s\n", r.URL.Path, err)
-		w.WriteHeader(http.StatusInternalServerError)
-		sendMetric(r.URL.Path,http.StatusInternalServerError, r.Method)
-		return
-	}
-	sendMetric(r.URL.Path,http.StatusOK, r.Method)
+	resp.Message = "success"
+	returnOk(w,r,resp)
 }
 
 func yumAction (w http.ResponseWriter, r *http.Request) {
 	vars := mux.Vars(r)
 	action := vars["action"]
 
+	resp := Response{}
+	resp.Namespace = string(r.URL.Path)
+
 	enabled, err := checkConfig(action)
 	if err != nil{
-		log.Printf("[ERROR] %s : %s\n", r.URL.Path, err)
-		w.WriteHeader(http.StatusNotFound)
-		sendMetric(r.URL.Path,http.StatusNotFound, r.Method)
+		returnBad(w,r,resp, err)
 		return
 	}
 
 	if !enabled{
-		log.Printf("[WARN] %s : %s", r.URL.Path, "API function disabled")
-		w.WriteHeader(http.StatusBadRequest)
-		sendMetric(r.URL.Path,http.StatusBadRequest, r.Method)
+		returnBad(w,r,resp, errors.New("API function disabled"))
 		return
 	}
 
 	if err := command(string("apt-get"), []string{action , "-y"}); err != nil{
-		log.Printf("[ERROR] %s : %s", r.URL.Path, err)
-		w.WriteHeader(http.StatusInternalServerError)
-		sendMetric(r.URL.Path,http.StatusInternalServerError, r.Method)
+		returnInternalError(w,r,resp, err)
 		return
 	}
 
-	outputs := Response{}
-	outputs.Namespace = string(r.URL.Path)
-	outputs.Message = "success"
-
-	w.Header().Set("Content-Type", "application/json; charset=UTF-8")
-
-	if err := json.NewEncoder(w).Encode(outputs); err != nil {
-		log.Printf("[ERROR] %s : %s\n", r.URL.Path, err)
-		w.WriteHeader(http.StatusInternalServerError)
-		sendMetric(r.URL.Path,http.StatusInternalServerError, r.Method)
-		return
-	}
-	sendMetric(r.URL.Path,http.StatusOK, r.Method)
+	resp.Message = "success"
+	returnOk(w,r,resp)
 }
 
 func serviceAction (w http.ResponseWriter, r *http.Request) {
@@ -109,217 +118,145 @@ func serviceAction (w http.ResponseWriter, r *http.Request) {
 	service := vars["service"]
 	action := vars["action"]
 
+	resp := Response{}
+	resp.Namespace = string(r.URL.Path)
+
 	enabled, err := checkConfig("service")
 	if err != nil{
-		log.Printf("[ERROR] %s : %s\n", r.URL.Path, err)
-		w.WriteHeader(http.StatusNotFound)
-		sendMetric(r.URL.Path,http.StatusNotFound, r.Method)
+		returnBad(w,r,resp, err)
 		return
 	}
 
 	if !enabled{
-		log.Printf("[WARN] %s : %s", r.URL.Path, "API function disabled")
-		w.WriteHeader(http.StatusBadRequest)
-		sendMetric(r.URL.Path,http.StatusBadRequest, r.Method)
+		returnBad(w,r,resp, errors.New("API function disabled"))
 		return
 	}
 
 	if err := validateServiceAction(action); err != nil{
-		log.Printf("[WARN] %s : %s", r.URL.Path, err)
-		w.WriteHeader(http.StatusBadRequest)
-		sendMetric(r.URL.Path,http.StatusBadRequest, r.Method)
+		returnBad(w,r,resp, err)
 		return
 	}
 
-	if err := command(string("service"), []string{service , action}); err != nil{
-		log.Printf("[ERROR] %s : %s", r.URL.Path, err)
-		w.WriteHeader(http.StatusInternalServerError)
-		sendMetric(r.URL.Path,http.StatusInternalServerError, r.Method)
+	if err := command(string("service"), []string{service, action}); err != nil{
+		returnInternalError(w,r,resp, err)
 		return
 	}
 
-	outputs := Response{}
-	outputs.Namespace = string(r.URL.Path)
-	outputs.Message = "success"
-
-	w.Header().Set("Content-Type", "application/json; charset=UTF-8")
-
-	if err := json.NewEncoder(w).Encode(outputs); err != nil {
-		log.Printf("[ERROR] %s : %s\n", r.URL.Path, err)
-		w.WriteHeader(http.StatusInternalServerError)
-		sendMetric(r.URL.Path,http.StatusInternalServerError, r.Method)
-		return
-	}
-	sendMetric(r.URL.Path,http.StatusOK, r.Method)
+	resp.Message = "success"
+	returnOk(w,r,resp)
 }
 
 func displayAction (w http.ResponseWriter, r *http.Request) {
 	vars := mux.Vars(r)
 	action := vars["action"]
 
+	resp := Response{}
+	resp.Namespace = string(r.URL.Path)
+
 	enabled, err := checkConfig("display")
 	if err != nil{
-		log.Printf("[ERROR] %s : %s\n", r.URL.Path, err)
-		w.WriteHeader(http.StatusNotFound)
-		sendMetric(r.URL.Path,http.StatusNotFound, r.Method)
+		returnBad(w,r,resp, err)
 		return
 	}
 
 	if !enabled{
-		log.Printf("[WARN] %s : %s", r.URL.Path, "API function disabled")
-		w.WriteHeader(http.StatusBadRequest)
-		sendMetric(r.URL.Path,http.StatusBadRequest, r.Method)
+		returnBad(w,r,resp, errors.New("API function disabled"))
 		return
 	}
 
 	newState, err := validateDisplayAction(action)
 	if err != nil{
-		log.Printf("[WARN] %s : %s", r.URL.Path, err)
-		w.WriteHeader(http.StatusBadRequest)
-		sendMetric(r.URL.Path,http.StatusBadRequest, r.Method)
+		returnBad(w,r,resp, err)
 		return
 	}
 
 	if err := command(string("vcgencmd"), []string{"display_power", newState}); err != nil{
-		log.Printf("[ERROR] %s : %s", r.URL.Path, err)
-		w.WriteHeader(http.StatusInternalServerError)
-		sendMetric(r.URL.Path,http.StatusInternalServerError, r.Method)
+		returnInternalError(w,r,resp, err)
 		return
 	}
 
-	outputs := Response{}
-	outputs.Namespace = string(r.URL.Path)
-	outputs.Message = "success"
-
-	w.Header().Set("Content-Type", "application/json; charset=UTF-8")
-
-	if err := json.NewEncoder(w).Encode(outputs); err != nil {
-		log.Printf("[ERROR] %s : %s\n", r.URL.Path, err)
-		w.WriteHeader(http.StatusInternalServerError)
-		sendMetric(r.URL.Path,http.StatusInternalServerError, r.Method)
-		return
-	}
-	sendMetric(r.URL.Path,http.StatusOK, r.Method)
+	resp.Message = "success"
+	returnOk(w,r,resp)
 }
 
 func gpioSwitch (w http.ResponseWriter, r *http.Request) {
 	vars := mux.Vars(r)
 	pin := vars["number"]
 
+	resp := Response{}
+	resp.Namespace = string(r.URL.Path)
+
+
 	enabled, err := checkConfig("gpio")
 	if err != nil{
-		log.Printf("[ERROR] %s : %s\n", r.URL.Path, err)
-		w.WriteHeader(http.StatusNotFound)
-		sendMetric(r.URL.Path,http.StatusNotFound, r.Method)
+		returnBad(w,r,resp, err)
 		return
 	}
 
 	if !enabled{
-		log.Printf("[WARN] %s : %s", r.URL.Path, "API function disabled")
-		w.WriteHeader(http.StatusBadRequest)
-		sendMetric(r.URL.Path,http.StatusBadRequest, r.Method)
+		returnBad(w,r,resp, errors.New("API function disabled"))
 		return
 	}
 
 	if err := gpioToggle(pin); err != nil{
-		log.Printf("[ERROR] %s : %s\n", r.URL.Path, err)
-		w.WriteHeader(http.StatusInternalServerError)
-		sendMetric(r.URL.Path,http.StatusInternalServerError, r.Method)
+		returnInternalError(w,r,resp, err)
 	}
 
-	outputs := Response{}
-	outputs.Message = "success"
-	outputs.Namespace = string(r.URL.Path)
-
-	w.Header().Set("Content-Type", "application/json; charset=UTF-8")
-
-	if err := json.NewEncoder(w).Encode(outputs); err != nil {
-		log.Printf("[ERROR] %s : %s\n", r.URL.Path, err)
-		w.WriteHeader(http.StatusInternalServerError)
-		sendMetric(r.URL.Path,http.StatusInternalServerError, r.Method)
-		return
-	}
-	sendMetric(r.URL.Path,http.StatusOK, r.Method)
+	resp.Message = "success"
+	returnOk(w,r,resp)
 }
 
 func gpioPullDown (w http.ResponseWriter, r *http.Request) {
 	vars := mux.Vars(r)
 	pin := vars["number"]
 
+	resp := Response{}
+	resp.Namespace = string(r.URL.Path)
+
 	enabled, err := checkConfig("gpio")
 	if err != nil{
-		log.Printf("[ERROR] %s : %s\n", r.URL.Path, err)
-		w.WriteHeader(http.StatusNotFound)
-		sendMetric(r.URL.Path,http.StatusNotFound, r.Method)
+		returnBad(w,r,resp, err)
 		return
 	}
 
 	if !enabled{
-		log.Printf("[WARN] %s : %s", r.URL.Path, "API function disabled")
-		w.WriteHeader(http.StatusBadRequest)
-		sendMetric(r.URL.Path,http.StatusBadRequest, r.Method)
+		returnBad(w,r,resp, errors.New("API function disabled"))
 		return
 	}
 
 	pinState, err := gpioDown(pin)
 	if err != nil{
-		log.Printf("[ERROR] %s : %s\n", r.URL.Path, err)
-		w.WriteHeader(http.StatusInternalServerError)
-		sendMetric(r.URL.Path,http.StatusInternalServerError, r.Method)
+		returnInternalError(w,r,resp, err)
 	}
 
-	outputs := Response{}
-	outputs.Message = strconv.Itoa(int(pinState)) //Returns the pin readout
-	outputs.Namespace = string(r.URL.Path)
-
-	w.Header().Set("Content-Type", "application/json; charset=UTF-8")
-
-	if err := json.NewEncoder(w).Encode(outputs); err != nil {
-		log.Printf("[ERROR] %s : %s\n", r.URL.Path, err)
-		w.WriteHeader(http.StatusInternalServerError)
-		sendMetric(r.URL.Path,http.StatusInternalServerError, r.Method)
-		return
-	}
-	sendMetric(r.URL.Path,http.StatusOK, r.Method)
+	resp.Message = strconv.Itoa(int(pinState)) //Returns the pin readout
+	returnOk(w,r,resp)
 }
 
 func gpioPullUp (w http.ResponseWriter, r *http.Request) {
 	vars := mux.Vars(r)
 	pin := vars["number"]
 
+	resp := Response{}
+	resp.Namespace = string(r.URL.Path)
+
 	enabled, err := checkConfig("gpio")
 	if err != nil{
-		log.Printf("[ERROR] %s : %s\n", r.URL.Path, err)
-		w.WriteHeader(http.StatusNotFound)
-		sendMetric(r.URL.Path,http.StatusNotFound, r.Method)
+		returnBad(w,r,resp, err)
 		return
 	}
 
 	if !enabled{
-		log.Printf("[WARN] %s : %s", r.URL.Path, "API function disabled")
-		w.WriteHeader(http.StatusBadRequest)
-		sendMetric(r.URL.Path,http.StatusBadRequest, r.Method)
+		returnBad(w,r,resp, errors.New("API function disabled"))
 		return
 	}
 
 	pinState, err := gpioUp(pin)
 	if err != nil{
-		log.Printf("[ERROR] %s : %s\n", r.URL.Path, err)
-		w.WriteHeader(http.StatusInternalServerError)
-		sendMetric(r.URL.Path,http.StatusInternalServerError, r.Method)
-	}
-
-	outputs := Response{}
-	outputs.Message = strconv.Itoa(int(pinState)) //Returns the pin readout
-	outputs.Namespace = string(r.URL.Path)
-
-	w.Header().Set("Content-Type", "application/json; charset=UTF-8")
-
-	if err := json.NewEncoder(w).Encode(outputs); err != nil {
-		log.Printf("[ERROR] %s : %s\n", r.URL.Path, err)
-		w.WriteHeader(http.StatusInternalServerError)
-		sendMetric(r.URL.Path,http.StatusInternalServerError, r.Method)
+		returnInternalError(w,r,resp, err)
 		return
 	}
-	sendMetric(r.URL.Path,http.StatusOK, r.Method)
+
+	resp.Message = strconv.Itoa(int(pinState)) //Returns the pin readout
+	returnOk(w,r,resp)
 }

--- a/httpHandlers_test.go
+++ b/httpHandlers_test.go
@@ -2,24 +2,237 @@ package main
 
 import (
 	"fmt"
+	"github.com/gorilla/mux"
 	"net/http/httptest"
 	"strings"
 	"testing"
 )
 
 func TestHttpGetStatus(t *testing.T){
-	fmt.Println("Testing GET /api/alive")
+	fmt.Println("***Testing GET /api/alive***")
 	w := httptest.NewRecorder()
 	r := httptest.NewRequest("GET", "/api/alive", nil)
 	r.Header.Set("Content-Type", "application/json; charset=UTF-8")
 
 	getStatus(w,r)
 
-	//have to trim whatespace - I don't know why this is happening yet...
+	if w.Result().StatusCode != 200{
+		t.Errorf("handler returned non-200 response")
+	}
+
+	//have to trim whatespace in response body string - I don't know why this is happening yet...
 	resp := strings.Replace(w.Body.String(), "\n", "", -1)
 	expected := `{"namespace":"/api/alive","message":"alive"}`
 	if resp != expected {
 		t.Errorf("handler returned unexpected body: got %v want %v",
 			w.Body.String(), expected)
+	}
+}
+
+//Expect a non-200 return code in the tests because the tests are expected to run in docker.
+func TestHttpPostPower(t *testing.T){
+	actions := []string{"reboot", "shutdown"}
+
+	fmt.Printf("***Testing POST /api/power/{%s}***\n", actions)
+
+	for _, a := range actions{
+		w := httptest.NewRecorder()
+
+		r := httptest.NewRequest("POST", "/api/power/"+a, nil)
+		r.Header.Set("Content-Type", "application/json; charset=UTF-8")
+		r = mux.SetURLVars(r, map[string]string{"action": a})
+
+		powerAction(w,r)
+
+		if w.Result().StatusCode != 500{
+			t.Errorf("handler returned non-500 response, are you testing in Docker???")
+		}
+
+		//have to trim whatespace in response body string - I don't know why this is happening yet...
+		resp := strings.Replace(w.Body.String(), "\n", "", -1)
+		expected := `{"namespace":"/api/power/`+a+`","message":"failed starting `+a+` [+1]: : exec: \"`+a+`\": executable file not found in $PATH"}`
+		if resp != expected {
+			t.Errorf("handler returned unexpected body: got %v want %v",
+				w.Body.String(), expected)
+		}
+	}
+}
+
+func TestHttpPostYum(t *testing.T){
+	actions := []string{"update"}
+
+	fmt.Printf("***Testing POST /api/apt/{%s}***\n",actions)
+
+	for _, a := range actions{
+		w := httptest.NewRecorder()
+
+		r := httptest.NewRequest("POST", "/api/apt/"+a, nil)
+		r.Header.Set("Content-Type", "application/json; charset=UTF-8")
+		r = mux.SetURLVars(r, map[string]string{"action": a})
+
+		yumAction(w,r)
+
+		if w.Result().StatusCode != 200{
+			t.Errorf("handler returned non-200 response, are you testing in Docker???")
+		}
+
+		//have to trim whatespace in response body string - I don't know why this is happening yet...
+		resp := strings.Replace(w.Body.String(), "\n", "", -1)
+		expected := `{"namespace":"/api/apt/`+a+`","message":"success"}`
+		if resp != expected {
+			t.Errorf("handler returned unexpected body: got %v want %v",
+				w.Body.String(), expected)
+		}
+	}
+}
+
+func TestHttpPostService(t *testing.T){
+	service := "redis"
+	actions := []string{"stop", "start", "restart"}
+
+	fmt.Printf("***Testing POST /api/service/{%s}/{%s}***\n",service, actions)
+
+	for _, a := range actions{
+		w := httptest.NewRecorder()
+
+		r := httptest.NewRequest("POST", "/api/service/"+service+"/"+a, nil)
+		r.Header.Set("Content-Type", "application/json; charset=UTF-8")
+		r = mux.SetURLVars(r, map[string]string{"action": a, "service": service})
+
+		serviceAction(w,r)
+
+		if w.Result().StatusCode != 200{
+			t.Errorf("handler returned non-200 response, are you testing in Docker???")
+		}
+
+		//have to trim whatespace in response body string - I don't know why this is happening yet...
+		resp := strings.Replace(w.Body.String(), "\n", "", -1)
+		expected := `{"namespace":"/api/service/`+service+`/`+a+`","message":"success"}`
+		if resp != expected {
+			t.Errorf("handler returned unexpected body: got %v want %v",
+				w.Body.String(), expected)
+		}
+	}
+}
+
+//Expect a non-200 return code in the tests because the tests are expected to run in docker.
+func TestHttpPostDisplay(t *testing.T){
+	actions := []string{"on", "off"}
+
+	fmt.Printf("***Testing POST /api/display/{%s}***\n", actions)
+
+	for _, a := range actions{
+		i := ""
+		if a == "on"{
+			i = "1"
+		} else {
+			i = "0"
+		}
+		w := httptest.NewRecorder()
+
+		r := httptest.NewRequest("POST", "/api/display/"+a, nil)
+		r.Header.Set("Content-Type", "application/json; charset=UTF-8")
+		r = mux.SetURLVars(r, map[string]string{"action": a})
+
+		displayAction(w,r)
+
+		if w.Result().StatusCode != 500{
+			t.Errorf("handler returned non-500 response, are you testing in Docker???")
+		}
+
+		//have to trim whatespace in response body string - I don't know why this is happening yet...
+		resp := strings.Replace(w.Body.String(), "\n", "", -1)
+		expected := `{"namespace":"/api/display/`+a+`","message":"failed starting vcgencmd [display_power `+i+`]: : exec: \"vcgencmd\": executable file not found in $PATH"}`
+		if resp != expected {
+			t.Errorf("handler returned unexpected body: got %v want %v",
+				w.Body.String(), expected)
+		}
+	}
+}
+
+//Expect a non-200 return code in the tests because the tests are expected to run in docker.
+func TestHttpPostPinUp(t *testing.T){
+	actions := []string{"1", "3"}
+
+	fmt.Printf("***Testing POST /api/pullup/{%s}***\n", actions)
+
+	for _, a := range actions{
+		w := httptest.NewRecorder()
+
+		r := httptest.NewRequest("POST", "/api/pullup/"+a, nil)
+		r.Header.Set("Content-Type", "application/json; charset=UTF-8")
+		r = mux.SetURLVars(r, map[string]string{"action": a})
+
+		gpioPullUp(w,r)
+
+		if w.Result().StatusCode != 500{
+			t.Errorf("handler returned non-500 response, are you testing in Docker???")
+		}
+
+		//have to trim whatespace in response body string - I don't know why this is happening yet...
+		resp := strings.Replace(w.Body.String(), "\n", "", -1)
+		expected := `{"namespace":"/api/pullup/`+a+`","message":"strconv.Atoi: parsing \"\": invalid syntax"}`
+		if resp != expected {
+			t.Errorf("handler returned unexpected body: got %v want %v",
+				w.Body.String(), expected)
+		}
+	}
+}
+
+//Expect a non-200 return code in the tests because the tests are expected to run in docker.
+func TestHttpPostPinDown(t *testing.T){
+	actions := []string{"1", "3"}
+
+	fmt.Printf("***Testing POST /api/pulldown/{%s}***\n", actions)
+
+	for _, a := range actions{
+		w := httptest.NewRecorder()
+
+		r := httptest.NewRequest("POST", "/api/pulldown/"+a, nil)
+		r.Header.Set("Content-Type", "application/json; charset=UTF-8")
+		r = mux.SetURLVars(r, map[string]string{"action": a})
+
+		gpioPullUp(w,r)
+
+		if w.Result().StatusCode != 500{
+			t.Errorf("handler returned non-500 response, are you testing in Docker???")
+		}
+
+		//have to trim whatespace in response body string - I don't know why this is happening yet...
+		resp := strings.Replace(w.Body.String(), "\n", "", -1)
+		expected := `{"namespace":"/api/pulldown/`+a+`","message":"strconv.Atoi: parsing \"\": invalid syntax"}`
+		if resp != expected {
+			t.Errorf("handler returned unexpected body: got %v want %v",
+				w.Body.String(), expected)
+		}
+	}
+}
+
+//Expect a non-200 return code in the tests because the tests are expected to run in docker.
+func TestHttpPostPinToggle(t *testing.T){
+	actions := []string{"1", "3"}
+
+	fmt.Printf("***Testing POST /api/toggle/{%s}***\n", actions)
+
+	for _, a := range actions{
+		w := httptest.NewRecorder()
+
+		r := httptest.NewRequest("POST", "/api/toggle/"+a, nil)
+		r.Header.Set("Content-Type", "application/json; charset=UTF-8")
+		r = mux.SetURLVars(r, map[string]string{"action": a})
+
+		gpioPullUp(w,r)
+
+		if w.Result().StatusCode != 500{
+			t.Errorf("handler returned non-500 response, are you testing in Docker???")
+		}
+
+		//have to trim whatespace in response body string - I don't know why this is happening yet...
+		resp := strings.Replace(w.Body.String(), "\n", "", -1)
+		expected := `{"namespace":"/api/toggle/`+a+`","message":"strconv.Atoi: parsing \"\": invalid syntax"}`
+		if resp != expected {
+			t.Errorf("handler returned unexpected body: got %v want %v",
+				w.Body.String(), expected)
+		}
 	}
 }

